BUG: f2py threads.h on MacOS 14.x may need a shim?

Downstream in SciPy (see: https://github.com/scipy/scipy/issues/21830), when using GNU toolchain version `14.2.0` from homebrew, I consistently see a SciPy build failure on MacOS ARM because of `scipy/linalg/_flapackmodule.c:90:10: fatal error: threads.h: No such file or directory`.

I made a temporary hack (below) to my local NumPy/`f2py` source that fixed the SciPy build issue. I'm not sure if that's exactly what we'd want to do, though [GNU docs on the matter](https://www.gnu.org/software/gnulib/manual/html_node/threads_002eh.html#:~:text=This%20header%20file%20is%20missing,%2C%20MSVC%2014%2C%20Android%209.0.) do suggest that `threads.h` is missing on MacOS 14. This part of `f2py` has caused pain a few times in the past per gh-19437 and gh-24761.

```diff
--- a/numpy/f2py/cfuncs.py
+++ b/numpy/f2py/cfuncs.py
@@ -558,7 +558,8 @@ def errmess(s: str) -> None:
       && (__STDC_VERSION__ >= 201112L) \\
       && !defined(__STDC_NO_THREADS__) \\
       && (!defined(__GLIBC__) || __GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ > 12)) \\
-      && !defined(NPY_OS_OPENBSD) && !defined(NPY_OS_HAIKU)
+      && !defined(NPY_OS_OPENBSD) && !defined(NPY_OS_HAIKU) \\
+      && !(defined(__APPLE__) && (__MAC_OS_X_VERSION_MAX_ALLOWED >= 140000 && __MAC_OS_X_VERSION_MAX_ALLOWED < 150000))
 /* __STDC_NO_THREADS__ was first defined in a maintenance release of glibc 2.12,
    see https://lists.gnu.org/archive/html/commit-hurd/2012-07/msg00180.html,
    so `!defined(__STDC_NO_THREADS__)` may give false positive for the existence
```

`threads.h` is also missing on earlier macOS, and the source erroneously assumes it is present due to C version.

```
[2/508] Compiling C object scipy/linalg/_interpolative.cpython-311-darwin.so.p/meson-generated__interpolativemodule.c.o
FAILED: scipy/linalg/_interpolative.cpython-311-darwin.so.p/meson-generated__interpolativemodule.c.o 
/opt/local/bin/gcc-mp-14 -Iscipy/linalg/_interpolative.cpython-311-darwin.so.p -Iscipy/linalg -I../scipy/linalg -I../../../../../../../../Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/numpy/core/include -I../../../../../../../../Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/numpy/f2py/src -I/opt/local/include/openblas -I/opt/local/Library/Frameworks/Python.framework/Versions/3.11/include/python3.11 -fvisibility=hidden -fdiagnostics-color=always -DNDEBUG -Wall -Winvalid-pch -std=c17 -O3 -Wno-unused-but-set-variable -Wno-unused-function -Wno-conversion -Wno-misleading-indentation -isysroot/ -MD -MQ scipy/linalg/_interpolative.cpython-311-darwin.so.p/meson-generated__interpolativemodule.c.o -MF scipy/linalg/_interpolative.cpython-311-darwin.so.p/meson-generated__interpolativemodule.c.o.d -o scipy/linalg/_interpolative.cpython-311-darwin.so.p/meson-generated__interpolativemodule.c.o -c scipy/linalg/_interpolative.cpython-311-darwin.so.p/_interpolativemodule.c
scipy/linalg/_interpolative.cpython-311-darwin.so.p/_interpolativemodule.c:96:10: fatal error: threads.h: No such file or directory
   96 | #include <threads.h>
      |          ^~~~~~~~~~~
compilation terminated.
[3/508] Compiling Fortran object scipy/linalg/_interpolative.cpython-311-darwin.so.p/meson-generated__interpolative-f2pywrappers.f.o
[4/508] Compiling C object scipy/linalg/_flapack.cpython-311-darwin.so.p/meson-generated_..__flapackmodule.c.o
FAILED: scipy/linalg/_flapack.cpython-311-darwin.so.p/meson-generated_..__flapackmodule.c.o 
/opt/local/bin/gcc-mp-14 -Iscipy/linalg/_flapack.cpython-311-darwin.so.p -Iscipy/linalg -I../scipy/linalg -I../../../../../../../../Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/numpy/core/include -I../../../../../../../../Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/numpy/f2py/src -I/opt/local/include/openblas -I/opt/local/Library/Frameworks/Python.framework/Versions/3.11/include/python3.11 -fvisibility=hidden -fdiagnostics-color=always -DNDEBUG -Wall -Winvalid-pch -std=c17 -O3 -Wno-unused-but-set-variable -Wno-unused-function -Wno-conversion -Wno-misleading-indentation -isysroot/ -Wno-empty-body -MD -MQ scipy/linalg/_flapack.cpython-311-darwin.so.p/meson-generated_..__flapackmodule.c.o -MF scipy/linalg/_flapack.cpython-311-darwin.so.p/meson-generated_..__flapackmodule.c.o.d -o scipy/linalg/_flapack.cpython-311-darwin.so.p/meson-generated_..__flapackmodule.c.o -c scipy/linalg/_flapackmodule.c
scipy/linalg/_flapackmodule.c:90:10: fatal error: threads.h: No such file or directory
   90 | #include <threads.h>
      |          ^~~~~~~~~~~
compilation terminated.
```

--- a/numpy/_core/meson.build
+++ b/numpy/_core/meson.build
@@ -245,8 +245,8 @@ endforeach
 
 # variable attributes tested via "int %s a" % attribute
 optional_variable_attributes = [
-  ['thread_local', 'HAVE_THREAD_LOCAL'],
-  ['_Thread_local', 'HAVE__THREAD_LOCAL'],
+  ['thread_local', 'HAVE_THREAD_LOCAL'],    # C23
+  ['_Thread_local', 'HAVE__THREAD_LOCAL'],  # C11/C17
   ['__thread', 'HAVE__THREAD'],
   ['__declspec(thread)', 'HAVE___DECLSPEC_THREAD_']
 ]

--- a/numpy/f2py/cfuncs.py
+++ b/numpy/f2py/cfuncs.py
@@ -553,7 +553,8 @@ cppmacros['OLDPYNUM'] = """
 # running a compile-time check (which we have no control over in generated
 # code used outside of NumPy) is hard. Therefore we support overriding this
 # via an external define - the f2py-using package can then use the same
-# compile-time checks as we use for `NPY_TLS` when building NumPy.
+# compile-time checks as we use for `NPY_TLS` when building NumPy (see
+# scipy#21860 for an example of that).
 #
 # __STDC_NO_THREADS__ should not be coupled to the availability of _Thread_local.
 # In case we get a bug report, guard it with __STDC_NO_THREADS__ after all.

--- a/numpy/f2py/cfuncs.py
+++ b/numpy/f2py/cfuncs.py
@@ -552,7 +552,7 @@ cppmacros['OLDPYNUM'] = """
 # Defining the correct value to indicate thread-local storage in C without
 # running a compile-time check (which we have no control over in generated
 # code used outside of NumPy) is hard. Therefore we support overriding this
-# via an external define - the f2py-using package can them use the same
+# via an external define - the f2py-using package can then use the same
 # compile-time checks as we use for `NPY_TLS` when building NumPy.
 #
 # __STDC_NO_THREADS__ should not be coupled to the availability of _Thread_local.

--- a/numpy/f2py/cfuncs.py
+++ b/numpy/f2py/cfuncs.py
@@ -548,24 +548,31 @@ cppmacros['OLDPYNUM'] = """
 #error You need to install NumPy version 0.13 or higher. See https://scipy.org/install.html
 #endif
 """
+
+# Defining the correct value to indicate thread-local storage in C without
+# running a compile-time check (which we have no control over in generated
+# code used outside of NumPy) is hard. Therefore we support overriding this
+# via an external define - the f2py-using package can them use the same
+# compile-time checks as we use for `NPY_TLS` when building NumPy.
+#
+# __STDC_NO_THREADS__ should not be coupled to the availability of _Thread_local.
+# In case we get a bug report, guard it with __STDC_NO_THREADS__ after all.
+#
+# `thread_local` has become a keyword in C23, but don't try to use that yet
+# (too new, doing so while C23 support is preliminary will likely cause more
+#  problems than it solves).
+#
+# Note: do not try to use `threads.h`, its availability is very low
+# *and* threads.h isn't actually used where `F2PY_THREAD_LOCAL_DECL` is
+# in the generated code. See gh-27718 for more details.
 cppmacros["F2PY_THREAD_LOCAL_DECL"] = """
 #ifndef F2PY_THREAD_LOCAL_DECL
 #if defined(_MSC_VER)
 #define F2PY_THREAD_LOCAL_DECL __declspec(thread)
 #elif defined(NPY_OS_MINGW)
 #define F2PY_THREAD_LOCAL_DECL __thread
-#elif defined(__STDC_VERSION__) \\
-      && (__STDC_VERSION__ >= 201112L) \\
-      && !defined(__STDC_NO_THREADS__) \\
-      && (!defined(__GLIBC__) || __GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ > 12)) \\
-      && !defined(NPY_OS_OPENBSD) && !defined(NPY_OS_HAIKU)
-/* __STDC_NO_THREADS__ was first defined in a maintenance release of glibc 2.12,
-   see https://lists.gnu.org/archive/html/commit-hurd/2012-07/msg00180.html,
-   so `!defined(__STDC_NO_THREADS__)` may give false positive for the existence
-   of `threads.h` when using an older release of glibc 2.12
-   See gh-19437 for details on OpenBSD */
-#include <threads.h>
-#define F2PY_THREAD_LOCAL_DECL thread_local
+#elif defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 201112L)
+#define F2PY_THREAD_LOCAL_DECL _Thread_local
 #elif defined(__GNUC__) \\
       && (__GNUC__ > 4 || (__GNUC__ == 4 && (__GNUC_MINOR__ >= 4)))
 #define F2PY_THREAD_LOCAL_DECL __thread

